pubsub/producer.go

```diff
@@ -0,0 +1,242 @@
+	if cfg.RedisURL == "" {
+		return nil, fmt.Errorf("redis url cannot be empty")
+	}
+	c, err := redisutil.RedisClientFromURL(cfg.RedisURL)
+	if err != nil {
+		return nil, err
+	}
+	return &Producer[Request, Response]{
+		id:       uuid.NewString(),
+		client:   c,
+		cfg:      cfg,
+		promises: make(map[string]*containers.Promise[Response]),
+	}, nil
+}
+
+func (p *Producer[Request, Response]) Start(ctx context.Context) {
+	p.StopWaiter.Start(ctx, p)
+	p.StopWaiter.CallIteratively(
+		func(ctx context.Context) time.Duration {
+			msgs, err := p.checkPending(ctx)
+			if err != nil {
+				log.Error("Checking pending messages", "error", err)
+				return p.cfg.CheckPendingInterval
+			}
+			if len(msgs) == 0 {
+				return p.cfg.CheckPendingInterval
+			}
+			acked := make(map[string]Request)
+			for _, msg := range msgs {
+				if _, err := p.client.XAck(ctx, p.cfg.RedisStream, p.cfg.RedisGroup, msg.ID).Result(); err != nil {
+					log.Error("ACKing message", "error", err)
+					continue
+				}
+				acked[msg.ID] = msg.Value
+			}
+			for k, v := range acked {
+				// Only re-insert messages that were removed the the pending list first.
+				_, err := p.reproduce(ctx, v, k)
+				if err != nil {
+					log.Error("Re-inserting pending messages with inactive consumers", "error", err)
+				}
+			}
+			return p.cfg.CheckPendingInterval
+		},
+	)
+	// Iteratively check whether result were returned for some queries.
+	p.StopWaiter.CallIteratively(func(ctx context.Context) time.Duration {
+		p.promisesLock.Lock()
+		defer p.promisesLock.Unlock()
+		for id, promise := range p.promises {
+			res, err := p.client.Get(ctx, id).Result()
+			if err != nil {
+				if errors.Is(err, redis.Nil) {
+					continue
+				}
+				log.Error("Error reading value in redis", "key", id, "error", err)
+			}
+			var tmp Response
+			val, err := tmp.Unmarshal([]byte(res))
+			if err != nil {
+				log.Error("Error unmarshaling", "value", res, "error", err)
+				continue
+			}
+			promise.Produce(val)
+			delete(p.promises, id)
+		}
+		return p.cfg.CheckResultInterval
+	})
+}
+
+// reproduce is used when Producer claims ownership on the pending
+// message that was sent to inactive consumer and reinserts it into the stream,
+// so that seamlessly return the answer in the same promise.
+func (p *Producer[Request, Response]) reproduce(ctx context.Context, value Request, oldKey string) (*containers.Promise[Response], error) {
+	id, err := p.client.XAdd(ctx, &redis.XAddArgs{
+		Stream: p.cfg.RedisStream,
+		Values: map[string]any{messageKey: value.Marshal()},
+	}).Result()
+	if err != nil {
+		return nil, fmt.Errorf("adding values to redis: %w", err)
+	}
+	p.promisesLock.Lock()
+	defer p.promisesLock.Unlock()
+	promise := p.promises[oldKey]
+	if oldKey == "" || promise == nil {
+		pr := containers.NewPromise[Response](nil)
+		promise = &pr
+	}
+	p.promises[id] = promise
```

Comment:
do you have a test covering that?

Reply:
(meaning the full cycle where a consumer stops updating it's liveliness and the workflow is moved)